File: app/src/main/java/com/juniperphoton/myerlist/widget/CateCircleView.java
package com.juniperphoton.myerlist.widget;


import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;

import com.juniperphoton.myerlist.R;

public class CateCircleView extends CircleView {
    private boolean mSelected;
    private Context mContext;

    private Paint mDarkPaint = new Paint();
    private Paint mPaint = new Paint();

    private int mRadius = 0;

    public CateCircleView(Context context, AttributeSet attrs) {
        super(context, attrs);
        mContext = context;
        setBackground(ContextCompat.getDrawable(context, R.drawable.ripple_borderless));
    }

    @Override
    public void setColor(int color) {
        super.setColor(color);
        mPaint.setColor(color);
        mDarkPaint.setColor(Color.argb(255, (int) (Color.red(color) * 0.4),
                (int) (Color.green(color) * 0.4),
                (int) (Color.blue(color) * 0.4)));
    }

    public void setSelected(final boolean selected) {
        if (mSelected == selected) {
            return;
        }
        mSelected = selected;
        int radius = (int) (getLayoutParams().width / 2 * 0.5);
        ValueAnimator valueAnimator = ValueAnimator.ofInt(selected ? 0 : radius, selected ? radius : 0);
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                mRadius = (int) animation.getAnimatedValue();
                invalidate();
            }
        });
        valueAnimator.setDuration(300);
        valueAnimator.start();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawCircle(getWidth() / 2, getHeight() / 2, getWidth() / 2, mDarkPaint);
        canvas.drawCircle(getWidth() / 2, getHeight() / 2, (int) (getWidth() / 2 * 0.8), mPaint);
        if (mSelected) {
            canvas.drawCircle(getWidth() / 2, getHeight() / 2, mRadius, mDarkPaint);
        }
    }
}

File: app/src/main/java/com/juniperphoton/myerlist/widget/CateCircleView.kt
package com.juniperphoton.myerlist.widget

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import com.juniperphoton.myerlist.R

class CateCircleView(ctx: Context, attrs: AttributeSet?) : CircleView(ctx, attrs) {
    private val darkPaint = Paint()
    private val paint = Paint()

    private var radius = 0

    override var color: Int = Color.WHITE
        set(value) {
            field = value
            paint.color = value
            darkPaint.color = Color.argb(255, (Color.red(value) * 0.4).toInt(),
                    (Color.green(value) * 0.4).toInt(),
                    (Color.blue(value) * 0.4).toInt())
            invalidate()
        }

    private var valueAnimator: ValueAnimator? = null

    var inSelected: Boolean = false
        set(newValue) {
            if (field == newValue) {
                return
            }
            field = newValue
            val value = (layoutParams.width / 2 * 0.5).toInt()
            valueAnimator = ValueAnimator.ofInt(if (newValue) 0 else value, if (newValue) value else 0)
            valueAnimator?.let {
                it.addUpdateListener { animation ->
                    radius = animation.animatedValue as Int
                    invalidate()
                }
                it.duration = 300
                it.start()
            }
        }

    init {
        background = ContextCompat.getDrawable(ctx, R.drawable.ripple_borderless)
        darkPaint.isDither = true
        darkPaint.isAntiAlias = true
        paint.isDither = true
        paint.isAntiAlias = true
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawCircle((width / 2).toFloat(), (height / 2).toFloat(), (width / 2).toFloat(), darkPaint)
        canvas.drawCircle((width / 2).toFloat(), (height / 2).toFloat(), (width / 2 * 0.8).toInt().toFloat(), paint)
        if (inSelected) {
            canvas.drawCircle((width / 2).toFloat(), (height / 2).toFloat(), radius.toFloat(), darkPaint)
        }
    }
}